### ISSUE TITLE
Hyperparameter-Feinjustierung f

### ISSUE THREAD

**Issue by tabee** [tabee/Reinforcement-Learning-Snake#28]
<html><head></head><body><h3><strong>Anleitung zur Hyperparameter-Feinjustierung für dein PPO-Training mit Snake</strong></h3>
<p>Um die optimale Performance aus deinem Modell herauszuholen, kannst du systematisch verschiedene <strong>Hyperparameter</strong> testen. Hier erfährst du, wie du das am effizientesten machst.</p>
<hr>
<h2><strong>1. Systematisches Hyperparameter-Tuning (Grid Search &amp; Random Search)</strong></h2>
<p>Es gibt zwei Hauptmethoden, um Hyperparameter zu optimieren:</p>
<h3>✅ <strong>1.1 Grid Search (Manuelles Testen fester Werte)</strong></h3>
<p>Hier werden verschiedene Kombinationen getestet, z. B.:</p>

learning_rate | n_steps | batch_size | clip_range | ent_coef | vf_coef
-- | -- | -- | -- | -- | --
0.0003 | 512 | 128 | 0.2 | 0.01 | 0.5
0.0005 | 1024 | 256 | 0.2 | 0.02 | 0.5
0.0001 | 2048 | 128 | 0.1 | 0.01 | 0.3

<p>Falls das Modell <strong>zu langsam lernt</strong>, <strong>erhöhe <code inline="">learning_rate</code> auf <code inline="">0.0005</code></strong>. Falls es instabil ist, reduziere es auf <code inline="">0.0001</code>.</p>
<hr>
<h2><strong>3. Optimale Trainingsstrategie für Snake</strong></h2>
<p>Die folgenden <strong>Trainingsstufen</strong> helfen, das Modell effizient zu trainieren:</p>
<h3><strong>Phase 1: Exploration (0–100k Timesteps)</strong></h3>
<ul>
<li>Starte mit <strong>höherem <code inline="">ent_coef=0.02</code></strong>, damit das Modell verschiedene Strategien ausprobiert.</li>
<li>Falls die Schlange zu oft stirbt, erhöhe <code inline="">clip_range=0.3</code>, damit sich die Policy langsamer ändert.</li>
<li>Verwende kleinere <code inline="">n_steps=512</code>, falls Episoden kurz sind.</li>
</ul>
<h3><strong>Phase 2: Stabilisierung (100k–500k Timesteps)</strong></h3>
<ul>
<li>Falls das Modell zu zufällig spielt, <strong>reduziere <code inline="">ent_coef=0.01</code></strong>.</li>
<li>Falls das Modell stark schwankt, reduziere <code inline="">learning_rate=0.0001</code>.</li>
<li>Falls die Performance stagniert, erhöhe <code inline="">batch_size=256</code>, um stabilere Updates zu haben.</li>
</ul>
<h3><strong>Phase 3: Optimierung (500k+ Timesteps)</strong></h3>
<ul>
<li>Falls der Reward nicht mehr steigt, teste <code inline="">learning_rate=0.00005</code>.</li>
<li>Falls das Modell <strong>zu vorhersehbar spielt</strong>, setze <code inline="">ent_coef=0.015</code> zur besseren Exploration.</li>
<li>Falls Training sehr langsam ist, erhöhe <code inline="">n_steps=1024</code> oder <code inline="">2048</code>.</li>
</ul>
<hr>
<h2><strong>4. Bestes Modell speichern &amp; vergleichen</strong></h2>
<p>Nach der Optimierung willst du das <strong>beste Modell</strong> behalten. Nutze dazu einen <code inline="">EvalCallback</code>:</p>
<pre><code class="language-python">from stable_baselines3.common.callbacks import EvalCallback

eval_callback = EvalCallback(
    env,
    best_model_save_path="./best_model/",
    log_path="./logs/",
    eval_freq=10000,  # Alle 10k Timesteps testen
    deterministic=True,
    render=False
)

model.learn(total_timesteps=500000, callback=eval_callback)
</code></pre>
<p>📌 <strong>Das beste Modell wird in <code inline="">./best_model/</code> gespeichert und kann später weiter trainiert werden!</strong></p>
<hr>
<h2><strong>5. Fazit: Schnellste Methode zur Optimierung</strong></h2>
<ol>
<li>
<p><strong>Falls du wenig Zeit hast:</strong></p>
<ul>
<li>Starte mit <strong>empfohlenen Standardwerten</strong> (<code inline="">n_steps=512</code>, <code inline="">batch_size=128</code>, <code inline="">learning_rate=0.0003</code>).</li>
<li>Trainiere für <strong>250k Timesteps</strong> und analysiere den Reward.</li>
</ul>
</li>
<li>
<p><strong>Falls du beste Performance willst:</strong></p>
<ul>
<li>Nutze <strong>Random Search (Optuna) für Hyperparameter-Optimierung</strong>.</li>
<li>Trainiere schrittweise (Exploration → Stabilisierung → Optimierung).</li>
<li>Überwache mit <strong>TensorBoard</strong> (<code inline="">ep_rew_mean</code>, <code inline="">policy_loss</code>, <code inline="">entropy_loss</code>).</li>
<li>Speichere das <strong>beste Modell mit <code inline="">EvalCallback</code></strong>.</li>
</ul>
</li>
</ol>
<hr>
<p>Mit dieser Strategie kannst du das <strong>effizienteste Modell für Snake</strong> finden, ohne unnötig viele Ressourcen zu verbrauchen. 🚀</p>
<p>Braucht du noch eine Anleitung zur <strong>Visualisierung oder Performance-Analyse</strong> nach dem Training? 😊</p></body></html>

**Comment by tabee** [comment]
Im Folgenden findest du meine Einschätzung, welche Konfiguration (Konfig) bei 100.000 Timesteps und welche bei 1.000.000 Timesteps für dein Snake-Game (trainiert mit PPO) vermutlich am besten funktioniert – natürlich immer unter der Annahme, dass das Spiel und dessen Dynamik eher überschaubar sind:

---

### **Für 100.000 Timesteps – Konfiguration: config0**

**Gründe:**

- **Hohe Lernrate (0.002):**  
  Eine vergleichsweise hohe Lernrate führt dazu, dass das Modell in kurzer Zeit größere Parameterupdates durchführt. Bei einem limitierten Trainingsbudget (100k Timesteps) ist es oft von Vorteil, schneller in die Richtung einer brauchbaren Lösung zu kommen – auch wenn dies das Risiko birgt, instabil zu werden. Für ein relativ simples Umfeld wie das Snake-Game kann diese Aggressivität helfen, rasch sinnvolle Muster zu erkennen.

- **Moderate n_steps (1024) und hohe Anzahl von Epochen (n_epochs = 10):**  
  Durch das Sammeln von 1024 Schritten pro Update und anschließendes mehrmaliges (10-faches) Durcharbeiten des gesammelten Batchs wird die begrenzte Datenmenge intensiv genutzt. Dies kann in einer frühen Trainingsphase von Vorteil sein, um schnell Verbesserungen zu erzielen, da das Modell so mehrfach auf den gleichen Erfahrungen "feinjustiert" wird.

- **Weitere Standardwerte (gamma, gae_lambda, clip_range, etc.):**  
  Diese Parameter sind in einem Bereich gewählt, der im PPO-Setting häufig gut funktioniert. Somit ist die Konfiguration insgesamt darauf ausgelegt, rasch erste brauchbare Policy-Anpassungen zu erzielen.

---

### **Für 1.000.000 Timesteps – Konfiguration: config3**

**Gründe:**

- **Niedrige Lernrate (0.0001):**  
  Bei einem langen Training ist es entscheidend, dass das Modell kontinuierlich und stabil lernt, ohne zu „überschießen“. Eine niedrige Lernrate sorgt dafür, dass die Policy-Updates kleiner und damit stabiler ausfallen. Dies minimiert die Gefahr, dass der Lernprozess durch zu große Schritte in instabile Bereiche gerät.

- **Hohe n_steps (2048):**  
  Ein größeres Sammeln an Erfahrungen vor jedem Update ermöglicht, dass die Berechnungen von Advantage-Schätzungen (und damit des Gradienten) robuster werden. Gerade bei 1 Mio. Timesteps, wo der Agent sehr viele Episoden erlebt, hilft ein größerer Batch dabei, mehr statistische Sicherheit in den Updates zu gewinnen.

- **Kleinerer clip_range (0.1):**  
  Ein konservativeres Clipping der Policy-Updates (im Vergleich zu 0.2 in anderen Konfigurationen) führt zu kleineren Änderungen pro Update-Schritt. Dies ist besonders bei langem Training sinnvoll, um das Risiko von sprunghaften und instabilen Policy-Veränderungen zu minimieren.

- **Angepasster vf_coef (0.3):**  
  Ein niedrigerer Wert für den Verlustkoeffizienten der Value-Funktion signalisiert, dass der Fokus etwas stärker auf das Policy-Learning gelegt wird. In einem längeren Training, wo sich sowohl die Policy als auch die Value-Funktion langsam verfeinern, kann dies helfen, Überanpassungen an den Value-Funktion-Schätzungen zu vermeiden.

---

### **Zusammenfassung**

- **100.000 Timesteps:**  
  **config0** ist vermutlich am besten geeignet, da der hohe Lernratenwert in Kombination mit häufigen und intensiven Updates (durch 10 Epochen) ein schnelles Auffinden einer brauchbaren Policy ermöglicht. Dies ist ideal, wenn du nur eine begrenzte Anzahl an Timesteps zur Verfügung hast.

- **1.000.000 Timesteps:**  
  **config3** dürfte bei einem langen Training die besten Ergebnisse liefern. Die sehr niedrige Lernrate und das größere n_steps sorgen für einen stabileren und feineren Lernprozess, der über die vielen Timesteps hinweg eine kontinuierliche Verbesserung ermöglicht, ohne in instabile Bereiche zu geraten.

Natürlich handelt es sich hierbei um eine theoretische Einschätzung – empirische Tests und eventuell kleinere Anpassungen könnten notwendig sein, um die optimale Performance in deinem konkreten Snake-Game zu erreichen.

**Comment by tabee** [comment]
Model: best_model_config0/best_model
 calculate_average_score 14.1578

Model: best_model_config1/best_model
 calculate_average_score 14.4493

Model: best_model_config2/best_model
 calculate_average_score 10.5918

Model: best_model_config3/best_model
 calculate_average_score 10.2492

Model: ppo_snake_config0
 calculate_average_score 13.0597

Model: ppo_snake_config1
 calculate_average_score 15.4045

Model: ppo_snake_config2
 calculate_average_score 16.3455

Model: ppo_snake_config3
 calculate_average_score 10.242
